### PR TITLE
Add early stopping

### PR DESCRIPTION
**Pull request by lruczu**

### CODE REVIEW

File: embeddings/task/lightning_task/lightning_task.py [hunk 1]
```diff
@@ -40,6 +42,8 @@ def fit(
         self.trainer = pl.Trainer(default_root_dir=str(self.output_path), **self.task_train_kwargs)
         try:
             self.trainer.fit(self.model, data)
+            self.current_epoch = self.trainer.current_epoch
+            self.best_epoch = self.current_epoch - self.trainer.early_stopping_callback.patience
```
Comment (ktagowski) [inline embeddings/task/lightning_task/lightning_task.py:46]: I think that it should be isolated from training. I would rewrite it as a separate callback, like early stopping and do not rely with other callback.
```python
BestEpochCallback(monitor="val/loss", mode="min")
```

More information on callbacks: https://pytorch-lightning.readthedocs.io/en/stable/extensions/callbacks.html

**Review by ktagowski — CHANGES_REQUESTED** [review]

File: embeddings/task/lightning_task/lightning_task.py (hunk 1 above)
Comment (AdrianSzymczak) [reply embeddings/task/lightning_task/lightning_task.py:46]: https://pytorch-lightning.readthedocs.io/en/stable/common/early_stopping.html

**Review by ktagowski — CHANGES_REQUESTED** [review]
BestEpochCallback refactor suggestion.

**Review by ktagowski — CHANGES_REQUESTED** [review]

File: embeddings/task/lightning_task/lightning_task.py [hunk 2]
```diff
@@ -24,20 +26,43 @@ def __init__(
...
     def fit(
         self,
         data: HuggingFaceDataModule,
     ) -> None:
         if not self.model:
             raise self.MODEL_UNDEFINED_EXCEPTION
-        self.trainer = pl.Trainer(default_root_dir=str(self.output_path), **self.task_train_kwargs)
+        self.trainer = pl.Trainer(
+            default_root_dir=str(self.output_path),
+            callbacks=[BestEpochCallback(), EarlyStopping(**self.early_stopping_kwargs)],
```
Comment (pedrito87) [inline embeddings/task/lightning_task/lightning_task.py:64]: It seems that we are enforcing the usage of these callbacks on the user, is it intended?

**Review by ktagowski — CHANGES_REQUESTED** [review]

File: embeddings/pipeline/lightning_classification.py [hunk 3]
```diff
@@ -22,6 +22,7 @@ class LightningClassificationPipeline(
     DEFAULT_TASK_MODEL_KWARGS = {"use_scheduler": True}
     DEFAULT_DATAMODULE_KWARGS = {"max_seq_length": None}
     DEFAULT_MODEL_CONFIG_KWARGS = {"classifier_dropout": None}
+    DEFAULT_EARLY_STOPPING_KWARGS = {"monitor": "val/Loss", "mode": "min", "patience": 1}
```
Comment (ktagowski) [inline embeddings/pipeline/lightning_classification.py:25]: Early stopping should be disabled if validation subset is not present (e.g. sequence labeling case)

**Review by ktagowski — DISMISSED** [review]

File: tests/test_callback_best_epoch.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    _ = pipeline.run()
+
+    # val/Loss epoch 0: 1.3314
+    # val/Loss epoch 1: 1.3253
+    # val/Loss epoch 2: 1.3278
+
+    np.testing.assert_equal(pipeline.model.task.best_epoch, 1)
+    np.testing.assert_almost_equal(
+        pipeline.model.task.best_validation_score, 1.325, decimal=pytest.decimal
+    )
```
Comment (asawczyn) [inline tests/test_callback_best_epoch.py:100]: Should it be epoch 1? According to the comment, the best epoch is 2

File: tests/test_callback_best_epoch.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    test_lightning_classification_pipeline(
+        lightning_classification_pipeline(
+            pipeline_kwargs(),
+            dataset_kwargs(),
+            datamodule_kwargs(),
+            task_train_kwargs(),
+            task_model_kwargs(),
+            TemporaryDirectory(),
+        )
+    )
```
Comment (asawczyn) [inline tests/test_callback_best_epoch.py:114]: remove it, tests will be detected automatically by pytest

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: tests/test_callback_best_epoch.py (hunk 4 above)
Comment (lruczu) [reply tests/test_callback_best_epoch.py:100]: It's ok. 1.3253 < 1.3278

**Review by asawczyn — APPROVED** [review]

**Review by ktagowski — APPROVED** [review]

**Pull request merged**